### PR TITLE
New Autonomy Layout

### PR DESCRIPTION
**Pull request by chachmu**
This adds the new system for organizing our autonomy. A very basic explanation is that individual tasks inherit from the BaseTask class and use state machines to complete their task. Multiple successive tasks are organized into different runs (prequal, finals, etc.) which inherit from the BaseRun class.

The BaseTask class provides some simple preprepared functionality to publish a debug string and an overlay image. I also added a small helper class that handles thresholding CV2 images so we don't have to recreate the wheel in every task that needs thresholding.

In the future it would be good to automate some more of the setup and state machine handling for the BaseTask class. We might also consider switching to StrEnums with auto() instead of the current type.

This PR also changes our simulated sonar to the sonoptix echo and adds a ping360 to the top of the vehicle.

### CODE REVIEW

File: pontus_autonomy/pontus_autonomy/base_run.py
Comment (Jeff300fang) [inline pontus_autonomy/pontus_autonomy/base_run.py:1]: nit: 

`task = task()` is a bit confusing. Without tracing `run_task`, one might be confused on what calling `task()` does. I suggest something like:

```
def run_task(self, Task):
      task = Task()
      ...
 ```
 or
```
def run_task(self, TaskClass):
      task = TaskClass()
      ...
 ```

File: pontus_autonomy/pontus_autonomy/prequalification_run.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+class PrequalificationRun(BaseRun):
+
+    def __init__(self):
+        super().__init__("prequalification_run")
+
+        print("Starting Prequalification Run")
+
+        # Gate Task
+        result = self.run_task(GateTask)
```
Comment (Jeff300fang) [inline pontus_autonomy/pontus_autonomy/prequalification_run.py:21]: Outside of prequal, how should information, such as the direction to circumnavigate the buoy, which is discovered during `self.run_task(GateTask)`, be passed back to the Run?

**Review by Jeff300fang — CHANGES_REQUESTED** [review]
Looks good.

File: pontus_autonomy/pontus_autonomy/helpers/cv_threshold.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        self.upper2 = upper2
+
+        self.masked = None
+
+    def mask_image(self, image):
+        self.mask = cv2.inRange(image, self.lower1, self.upper1)
+        if self.lower2 is not None and self.upper2 is not None:
+            self.mask += cv2.inRange(image, self.lower2, self.upper2)
+
+        return self.mask
```
[the remark this answers is not in the document]
Comment (chachmu) [reply pontus_autonomy/pontus_autonomy/helpers/cv_threshold.py:19]: I guess you could try to generalize it, but you shouldn't ever have more than 2 sets of values anyway. The only reason for multiple masks was to handle what happens when you have a threshold that you want to wrap around the limits of the hsv values (for example if you want to detect red you need to look for hues from 170-10 degrees)

File: pontus_autonomy/pontus_autonomy/helpers/cv_threshold.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class CVThreshold:
+    def __init__(self, lower1, upper1, lower2 = None, upper2 = None):
+        self.lower1 = lower1
+        self.upper1 = upper1
+
+        self.lower2 = lower2
+        self.upper2 = upper2
+
+        self.masked = None
```
[the remark this answers is not in the document]
Comment (chachmu) [reply pontus_autonomy/pontus_autonomy/helpers/cv_threshold.py:12]: No, `mask` and `masked` are different, I should probably add a comment to the code. `mask` is the black and white mask that tells you which parts of the image you want, `masked` is what you get after multiplying the mask with the image to get black everywhere except the areas that you were trying to select and color in those regions.

**Review by Jeff300fang — APPROVED** [review]

**Pull request merged**